File: src/spirite/base/image_data/images/drawer/IImageDrawer.java
package spirite.base.image_data.images.drawer;

import spirite.base.brains.ToolsetManager.ColorChangeScopes;
import spirite.base.image_data.ImageWorkspace.BuildingImageData;
import spirite.base.image_data.SelectionEngine.BuiltSelection;
import spirite.base.image_data.UndoEngine.ImageAction;
import spirite.base.pen.PenTraits.PenState;
import spirite.base.pen.StrokeEngine;
import spirite.base.util.glmath.MatTrans;

public interface IImageDrawer {

	public abstract class MaskedImageAction extends ImageAction {
		protected final BuiltSelection mask;

		protected MaskedImageAction(BuildingImageData data, BuiltSelection mask) {
			super(data);
			this.mask = mask;
		}
	}
	
	// Modules, an Image Drawer may implement these or they may not.  Not implementing them means
	//	that the Drawer is incapable of performing these draw actions (e.g. because it doesn't 
	//	make sense for the data type).
	public interface IStrokeModule {
		//public boolean strokeIsDrawing();	// SHOULD be true iff getStrokeEngine = null
		public StrokeEngine getStrokeEngine();
		public boolean canDoStroke( StrokeEngine.StrokeParams params);	// EG: some Drawers might be able to erase, but not draw
		public boolean startStroke( StrokeEngine.StrokeParams params, PenState ps);
		public void stepStroke( PenState ps);
		public void endStroke();
	}
	
	public interface IClearModule {
		public void clear();
	}
	
	public interface IFillModule {
		public boolean fill( int x, int y, int color, BuildingImageData _data);
	}
	
	public interface IFlipModule {
		public void flip( boolean horizontal);
	}
	
	public interface IColorChangeModule {
		public void changeColor(  int from, int to, ColorChangeScopes scope, int mode);
	}
	
	public interface IInvertModule {
		public void invert();
	}
	
	public interface ITransformModule {
		public void transform( MatTrans trans);
	}
	
	public interface IWeightEraserModule {
		public void startWeightErase();
		public void endWeightErase();
		public void weightErase( float x, float y, float w);
	}
	
	public interface IMagneticFillModule {
		public void startMagneticFill();
		public void endMagneticFill( int color);
		public void anchorPoints(float x, float y, float r, boolean locked, boolean relooping);
		public void erasePoints( float x, float y, float r);
		public float[] getMagFillXs();
		public float[] getMagFillYs();
	}
}

File: src/main/kotlin/spirite/base/imageData/drawer/IImageDrawer.kt
package spirite.base.imageData.drawer

import rb.glow.color.Color
import rb.glow.color.SColor
import rb.vectrix.linear.ITransformF
import rb.vectrix.linear.ImmutableTransformF
import spirite.base.brains.toolset.ColorChangeMode
import spirite.base.brains.toolset.MagneticFillMode
import spirite.base.imageData.deformation.IDeformation
import spirite.base.imageData.mediums.ArrangedMediumData
import spirite.base.imageData.mediums.BuiltMediumData
import spirite.base.imageData.selection.ILiftedData
import spirite.base.imageData.selection.ISelectionExtra
import spirite.base.imageData.selection.Selection
import spirite.base.imageData.undo.IUndoEngine
import spirite.base.imageData.undo.ImageAction
import spirite.base.pen.PenState
import spirite.base.pen.stroke.StrokeParams
import spirite.base.util.linear.Rect


fun IUndoEngine.performMaskedImageAction(description: String, arranged: ArrangedMediumData, mask: Selection?, action: (BuiltMediumData, Selection?)->Any? )
{
    performAndStore(object: ImageAction(arranged) {
        override val description: String get() = description

        override fun performImageAction(built: BuiltMediumData) {
            action(built, mask)
        }
    })
}

interface IImageDrawer {
    // Modules: an Image Drawer may implement these or they may not.  Not implementing them means
    //	that the Drawer is incapable of performing these draw actions (e.g. because it doesn't
    //	make sense for the data type).

    interface IStrokeModule {
        // EG: some Drawers might be able to erase, but not draw
        fun canDoStroke(method: StrokeParams.Method): Boolean
        fun startStroke(params: StrokeParams, ps: PenState): Boolean
        fun stepStroke(ps: PenState)
        fun endStroke()
    }

    interface IClearModule {
        fun clear()
    }

    interface IFillModule {
        fun fill(x: Int, y: Int, color: Color): Boolean
    }

    interface IFlipModule {
        // Note: Theoretically, a Drawer could implement Flipping without having to implement full matrix transformation
        fun flip(horizontal: Boolean)
    }
    interface ITransformModule : IFlipModule {
        fun transform(trans: ITransformF, centered: Boolean = true)

        fun startManipulatingTransform() : Rect?
        fun stepManipulatingTransform()
        fun endManipulatingTransform()

        override fun flip(horizontal: Boolean) {
            transform(when( horizontal) {
                true -> ImmutableTransformF.Scale(-1f,1f)
                false -> ImmutableTransformF.Scale(1f, -1f)
            })
        }
    }

    interface IColorChangeModule {
        fun changeColor(from: Color, to: Color, mode: ColorChangeMode)
    }

    interface IInvertModule {
        fun invert()
    }


//    interface IWeightEraserModule {
//        fun startWeightErase(precise: Boolean)
//        fun endWeightErase()
//        fun weightErase(xi: Float, yi: Float, wf: Float)
//    }
//
    interface IMagneticFillModule {
        val magFillXs: FloatArray
        val magFillYs: FloatArray
        fun startMagneticFill()
        fun endMagneticFill(color: SColor, mode: MagneticFillMode)
        fun anchorPoints(x: Float, y: Float, r: Float, locked: Boolean, relooping: Boolean)
        fun erasePoints(x: Float, y: Float, r: Float)
    }

    interface IMagneticEraseModule {
        fun erase( x: Float, y: Float)
    }

    interface ILiftSelectionModule {
        fun liftSelection(selection: Selection, clearLifted: Boolean = true): ILiftedData?
        fun getSelectionExtra(selection: Selection) : ISelectionExtra? = null
    }

    interface IAnchorLiftModule {
        fun acceptsLifted(lifted: ILiftedData): Boolean
        fun anchorLifted(lifted: ILiftedData, tMediumToLifted: ITransformF?)
    }

    interface IDeformDrawer {
        fun deform(deformation: IDeformation)
    }
//
//    interface IBoneDrawer {
//        fun contort(bone: BaseBone, to: Interpolator2D)
//    }
//
//    interface IPuppetBoneDrawer {
//        fun grabBone(xi: Int, yi: Int, width: Float): BaseBone
//        fun makeBone(x1: Float, y1: Float, x2: Float, y2: Float)
//    }
}

object NillImageDrawer : IImageDrawer